code cleanups via IntelliJ IDEA inspections

jlawrynowicz opened the pull request:
I've applied some automatic inspections provided via IntelliJ IDEA to fix some minor code issues. There shouldn't be functional changes of any kind included here. There aren't any JDK8 or above features used.

Unfortunately I messed up the formatting and fixed it in the squashed commit. If you think any type of changes should be reverted, I'll recreate this squashed commit without given change.

I wrote the code, it's mine, and I'm contributing it to H2 for distribution multiple-licensed under the MPL 2.0, and the EPL 1.0 (http://h2database.com/html/license.html).

See commit message for commits included in squash.

jlawrynowicz committed:
cleanup - explicit Object extending

jlawrynowicz committed:
cleanup - avoid checking for null before instanceof

jlawrynowicz committed:
cleanup - use contains instead of indexOf

jlawrynowicz committed:
cleanup - avoid unnecessary boxing

jlawrynowicz committed:
cleanup - redundant final on static methods

jlawrynowicz committed:
performance - redundant array creation for vararg calls

jlawrynowicz committed:
performance - use Colections.singletonList instead of list when possible

jlawrynowicz committed:
performance - avoid unnecessary String.substring calls

jlawrynowicz committed:
performance - duplicated delimiters in StringTokenizer

jlawrynowicz committed:
performance - do not check for string ==  when equals will be called

equals implementation for String begins with  == equality check

jlawrynowicz committed:
cleanup - use + instead of String.concat

jlawrynowicz committed:
cleanup - redundant new for constant array creation

jlawrynowicz committed:
cleanup - missorted identifiers

katzyn commented on h2/src/main/org/h2/value/DataType.java:
```diff
@@ -1335,19 +1328,19 @@ public static Object getDefaultForPrimitiveType(Class<?> clazz) {
+            return (float) 0;
         } else if (clazz == Double.TYPE) {
-            return Double.valueOf(0);
+            return (double) 0;
```
Why `0L` is used for `long`, but `0f` and `0d` aren't used for `float` and `double`?

jlawrynowicz replied:
I'll fix it. Would you prefer a rebase/amend or a new commit and merges?

katzyn commented on h2/src/main/org/h2/util/ParserUtil.java:
```diff
@@ -93,12 +93,13 @@ public static int getSaveTokenType(String s, boolean functionsAsKeywords) {
+                case "CONSTRAINT":
+                    return KEYWORD;
+                case "CROSS":
+                    return KEYWORD;
```
Actulally `switch` may be slower here, it's better to run JMH tests before such changes.

jlawrynowicz replied:
do you have any JMH performance test cases for H2 I could use to check for potential regressions I should I prepare some myself?

If not I'll prepare some JMH for this method only.

katzyn replied:
JMH is not for testing overall performance of large projects. It's for testing small methods. But to perform such test for such pieces of code you should ensure that you use a brand new string on each execution. Constant string are handled in faster way in `switch` because they already knows their hash code and result with them will be incorrect.

I run some tests with this specific part with `CONSTRAINT` and `CROSS` and `switch` is slower than a pair of `equals()` by a 30% on Java 9. This was expected. `Parser` is a performance-critical part of H2, so this slowdown is not good.

jlawrynowicz replied:
I prepared a JMH microbenchmark: https://github.com/jlawrynowicz/jmh-switch-microbenchmark

You were actually spot on - both on JDK7 and JDK9 version with first switch on character and then if/else was actually slightly faster then current version, one I suggested here and even simplified one (only one switch) and pure if/else implementation.

Weirdly enough, I also tested using static hashMap which was actually much faster (~49% for openjdk7, ~45% on oracle jdk9). 

I'll revert all the switch changes as they don't seem to be such a perf win as I thought they were, but perhaps we could look into the hashMap numbers - perhaps somethings too unrealistic with my benchmark.

```
# tested on linux 4.15.9, CPU: i7-4710HQ

# OpenJDK Runtime Environment (IcedTea 2.6.13) (Arch Linux build 7.u171_2.6.13-1-x86_64)
Benchmark                               Mode  Cnt   Score   Error  Units
JmhMicrobenchmark.hashMap              thrpt   10  58.551 ± 1.023  ops/s
JmhMicrobenchmark.if_else              thrpt   10  36.468 ± 0.313  ops/s
JmhMicrobenchmark.switch_nested        thrpt   10  38.419 ± 0.687  ops/s
JmhMicrobenchmark.switch_simplified    thrpt   10  37.635 ± 0.699  ops/s
JmhMicrobenchmark.switch_with_if_else  thrpt   10  39.302 ± 0.728  ops/s

# Java(TM) SE Runtime Environment (build 9.0.4+11)
Benchmark                               Mode  Cnt   Score   Error  Units
JmhMicrobenchmark.hashMap              thrpt   10  56.675 ± 0.400  ops/s
JmhMicrobenchmark.if_else              thrpt   10  36.036 ± 0.620  ops/s
JmhMicrobenchmark.switch_nested        thrpt   10  37.570 ± 0.613  ops/s
JmhMicrobenchmark.switch_simplified    thrpt   10  35.304 ± 0.601  ops/s
JmhMicrobenchmark.switch_with_if_else  thrpt   10  39.118 ± 0.566  ops/s
```

katzyn replied:
It's not easy to test hash-based implementations (including switch with strings) correctly.

`String.hashCode()` may be slow enough on the first invocation. There is a big difference between newly allocated strings and old string that already has computed hash code. But `ParserUtil` may got a new string without a hash code and may got a string from cache with known hash code and proportions of such variants are unknown.

You should pass at least some strings that are allocated with `new String(oldString.toCharArray())` or similar methods. `new String(oldString)` copies hash code from the old string and should not be used due to this reason. But allocation of such new strings should be excluded from the testing times. So proper test will be complex enough.

Also different input values are needed including custom strings that are not keywords.

From my point of view there is no sense to spend more time for such tests. It will take a lot of time and looks like there is no way to achieve a really significant performance win.

jlawrynowicz replied:
I reran the benchmark with worst case scenario of all strings newly created (should be even as all tests have the same number of new String(s.toCharArray()) calls and the results are again in favor of current implementation. It goes to show how well optimized it already was.

I guess without realistic, synthetic benchmark it's hard to think about optimizing it. I'd believe that due to java's string cache most strings would already have their hash code precalculated, so HashMap approach could be faster, but I have no hard data to back it up at the moment. I'll stop looking into it.

I already crunched the numbers, so I might as well post them for someone perhaps:
```
# tested on linux 4.15.9, CPU: i7-4710HQ

# OpenJDK Runtime Environment (IcedTea 2.6.13) (Arch Linux build 7.u171_2.6.13-1-x86_64)
Benchmark                               Mode  Cnt   Score   Error  Units
JmhMicrobenchmark.hashMap              thrpt   10  17.648 ± 0.536  ops/s
JmhMicrobenchmark.if_else              thrpt   10  20.526 ± 0.169  ops/s
JmhMicrobenchmark.switch_nested        thrpt   10  20.972 ± 0.167  ops/s
JmhMicrobenchmark.switch_simplified    thrpt   10  19.414 ± 0.191  ops/s
JmhMicrobenchmark.switch_with_if_else  thrpt   10  21.951 ± 0.192  ops/s

# Java(TM) SE Runtime Environment (build 9.0.4+11)
Benchmark                               Mode  Cnt   Score   Error  Units
JmhMicrobenchmark.hashMap              thrpt   10  17.202 ± 0.210  ops/s
JmhMicrobenchmark.if_else              thrpt   10  20.604 ± 0.148  ops/s
JmhMicrobenchmark.switch_nested        thrpt   10  19.990 ± 0.115  ops/s
JmhMicrobenchmark.switch_simplified    thrpt   10  18.297 ± 0.086  ops/s
JmhMicrobenchmark.switch_with_if_else  thrpt   10  20.790 ± 0.248  ops/s
```

katzyn commented on h2/src/main/org/h2/mvstore/DataUtils.java:
```diff
@@ -941,8 +941,8 @@ public static int getErrorCode(String m) {
-    public static long readHexLong(Map<String, ? extends Object> map,
-            String key, long defaultValue) {
+    public static long readHexLong(Map<String, ?> map,
+                                   String key, long defaultValue) {
```
Do not use such indentation, it will break build process of H2. Just add 8 more spaces for wrapped lines, and 4 spaces in other cases. Adjust code formatting settings in your IDE for H2 if it has such options.

Line length limit for code is 120 characters now (80 for comments), so line break here is not required (and not forbidden).

katzyn commented on h2/src/main/org/h2/mvstore/DataUtils.java:
```diff
@@ -1006,8 +1006,8 @@ public static int parseHexInt(String x) {
-    public static int readHexInt(HashMap<String, ? extends Object> map,
-            String key, int defaultValue) {
+    public static int readHexInt(HashMap<String, ?> map,
+                                 String key, int defaultValue) {
```
The same here.

katzyn commented on h2/src/main/org/h2/index/PageDataNode.java:
```diff
@@ -172,7 +172,7 @@ PageData split(int splitPoint) {
         PageDataNode p2 = PageDataNode.create(index, newPageId, parentPageId);
         int firstChild = childPageIds[splitPoint];
-        for (int i = splitPoint; i < entryCount;) {
+        for (; splitPoint < entryCount; ) {
```
`while` should be used instead of such `for`.

katzyn commented on h2/src/main/org/h2/index/PageDataLeaf.java:
```diff
@@ -370,7 +370,7 @@ int getEntryCount() {
         int newPageId = index.getPageStore().allocatePage();
         PageDataLeaf p2 = PageDataLeaf.create(index, newPageId, parentPageId);
-        for (int i = splitPoint; i < entryCount;) {
+        for (; splitPoint < entryCount; ) {
```
`while`

katzyn commented on h2/src/main/org/h2/index/IndexCondition.java:
```diff
@@ -155,7 +155,7 @@ public Value getCurrentValue(Session session) {
             valueSet.add(v);
         }
-        Value[] array = valueSet.toArray(new Value[valueSet.size()]);
+        Value[] array = valueSet.toArray(new Value[0]);
```
Please, revert it. `ArrayList.toArray(new T[0])` works faster than `ArrayList.toArray(new T[list.size()]`), but other collections are usually not. I checked all collections some time ago with JMH tests with different Java versions and decided to use 0 only for `ArrayList` based on results of these tests.

jlawrynowicz replied:
That's interesting - I'll revert it. 

BTW do you have those JMHs somewhere on github?

katzyn replied:
No, I had a lot of such tests, but I didn't publish code anywhere and I probably already overwrote or deleted them, I can't find them now.

You can check this article, it has some samples and a lot of results:
https://shipilev.net/blog/2016/arrays-wisdom-ancients/

But notice that you actually perform this change in a more complex collections, like key set views of maps, they are not implemented in the same way as `ArrayList.toArray(T[])`. They are not tested in that article, so additional testing is required.

Or you may just leave this code alone.

katzyn commented on h2/src/main/org/h2/engine/Session.java:
```diff
@@ -352,7 +352,7 @@ public Value getVariable(String name) {
             return new String[0];
         }
-        return variables.keySet().toArray(new String[variables.size()]);
+        return variables.keySet().toArray(new String[0]);
```
The same here.

katzyn commented on h2/src/main/org/h2/engine/Session.java:
```diff
@@ -833,7 +831,7 @@ public void rollbackTo(Savepoint savepoint, boolean trimToSize) {
         }
         if (savepoints != null) {
-            String[] names = savepoints.keySet().toArray(new String[savepoints.size()]);
+            String[] names = savepoints.keySet().toArray(new String[0]);
```
And here. Revert all such changes with such collections.

katzyn commented on h2/src/main/org/h2/command/dml/Set.java:
```diff
@@ -165,15 +165,15 @@ public int update() {
+                    newMode = CompareMode.getInstance(currentMode.getName(), currentMode.getStrength(), true);
+                    break;
+                default:
+                    throw DbException.getInvalidValueException("BINARY_COLLATION", stringValue);
```
`switch` here is also probably works slower, but this should be checked.

katzyn commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -4356,20 +4340,20 @@ private String readCommentIf() {
 
     private Column parseColumnWithType(String columnName) {
-        String original = currentToken;
+        StringBuilder original = new StringBuilder(currentToken);
```
It looks like this optimization may have negative overall effect, because in the most cases original string is not concatenated with different extensions.

katzyn reviewed (commented):
Many changes looks good, but this pull request is to huge to review and contains one large commit with too different and unrelated changes. It's a largest problem here. It's better to send smaller pull requests one per each optimization or at least add each kind of optimization in a separate commit.

Also many optimizations definitely have negative effect, like `toArray(new T[0])` on collections that do not like it (`ArrayList` is different and optimized well).

Some optimizations with strings have unknown effect on performance and probably should be tested, because such changes may have positive effect and may have negative effect.

jlawrynowicz commented:
Thanks for the notes, I'll prepare some JMH test cases for switch performance and will remove changes according to your comments.

As I wrote in the first message, I have all the changes as separate commits, but unfortunately I messed up the formatting and decided against fixing formatting commit by commit.

Would you prefer me to add fixes in separate commits or rebase/amend the commit to keep the history simpler?

katzyn commented:
This is too huge to review, I'm not sure that I checked everything.

Just perform each kind of change in a separate commit and send a new pull request with them all, or you can send each change in a separate pull request.

But avoid changes that a definitely not good, such as changes with `toArray()` for key set views and similar collections or switches where a few of comparison operations beat them in performance.

katzyn commented on h2/src/main/org/h2/bnf/context/DbSchema.java:
```diff
@@ -71,11 +71,8 @@
-            isSystem = false;
-        }
+        } else
+            isSystem = contents.isDerby() && name.startsWith("SYS");
```
I also don't think that this change and similar to it are good. Because they introduce mixed logic that complicates reading of sources. If you want change this method to save a few bytes in compiled bytecode, it's better to reorganize all these if-else branches into something like
```Java
isSystem =
    // comment1
    (condition1)
    // comment2
    || (condition2 && condition3)
    // comment3
    || (condition4 && (condition5)
..... 
```
If this is not possible in some case due to complex conditions, it's better to leave this code as is.

jlawrynowicz replied:
is the current version alright? I can't see the mixed logic here - it's just avoiding using additional if.

katzyn replied:
The problem here is that with this change case with Derby looks like “default” / “other” case. But this condition is not special and should be treated like conditions above it. But this is not clear for casual reader.

This only reduces bytecode size by two instructions and probably saves nothing after JIT compilation with C2. Readability is more important than this.

But the whole if-else if-else operator may be reorganized to avoid all these `= true` and `= false`, but only with proper line breaks and indentation to preserve readability. This reduces size of bytecode significantly. But this also is not necessary because this constructor is not performance-critical.

jlawrynowicz replied:
I applied your suggestion and removed inlined all ifs.

katzyn commented on h2/src/main/org/h2/value/Value.java:
```diff
@@ -429,55 +429,55 @@ public static void clearCache() {
 
     public long getLong() {
-        return ((ValueLong) convertTo(Value.LONG)).getLong();
+        return convertTo(Value.LONG).getLong();
```
All these typecasts look like a guards against invalid results of `convertTo()`.

After this change if result of `convertTo()` does not override corresponding getter behavior will be changed. We got `StackOverflowError` instead of `ClassCastException`. If some code expects possible `Exception`, but not expects possible `Error` this may lead to a problem.

Hypothetically we should not have such broken `convertTo()` implementations, but anyway it looks like this change needs an additional analysis and review.

katzyn commented on h2/src/main/org/h2/value/ValueEnum.java:
```diff
@@ -118,15 +118,15 @@ private static String sanitize(final String label) {
-        result += ")";
-        return result;
+        result.append(")");
+        return result.toString();
```
Usually `StatementBuilder` is used in H2 codebase for such operations, because in has `appendExceptFirst()` that may be used for `", "`.

katzyn commented on h2/src/main/org/h2/value/ValueLob.java:
```diff
@@ -297,23 +297,21 @@ private void createFromReader(char[] buff, int len, Reader in,
         int f = objectId % SysProperties.LOB_FILES_PER_DIRECTORY;
         if (f > 0) {
-            name = SysProperties.FILE_SEPARATOR + objectId;
+            name = new StringBuilder(SysProperties.FILE_SEPARATOR + objectId);
```
Why concatenation is used inside `new StringBuilder()`?

jlawrynowicz commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -3577,21 +3569,17 @@ private boolean equalsToken(String a, String b) {
-            return true;
-        }
-        return false;
+        } else
```
are the braces for single line if/else required or can it be left as it is?

jlawrynowicz replied:
nvm, I pushed the formatting fix

jlawrynowicz commented:
I removed changes you didn't like - toArray, switch, unnecessary casts, adding StringBuilder and also changed all incomplete for loops into while loops.

Unfortunately I did not split the commits into smaller ones, as it would be really time consuming (my non squashed commits have formatting broken). If it's required, I can attempt to do it, but I'd rather just remove the cleanups you don't like and squash again.

grandinj commented:
no thanks, if anything goes wrong, I don't want to bisect back through a 122 file patch, please split this up

jlawrynowicz committed:
cleanup - unnecessary semicolons, static keyword

jlawrynowicz committed:
cleanup - use conditions for loops instead of breaks

jlawrynowicz committed:
performance - use addAll instead of manually adding each element

jlawrynowicz committed:
performance - use initialCapacity for HashMap

jlawrynowicz committed:
cleanup - redundant continue statements

jlawrynowicz committed:
cleanup - redundant return statements

jlawrynowicz committed:
performance - fix passing concatenation to StringBuilder append

jlawrynowicz committed:
cleanup - redundant local variables

jlawrynowicz committed:
cleanup - default in switch should be after other cases

jlawrynowicz committed:
cleanup - use while loops instead of incomplete for loops

jlawrynowicz commented:
Bisecting is a fair point, I did not think about that. I split the commit, feel free to ask for any other changes.

jlawrynowicz committed:
formatting - add braces to control statements

jlawrynowicz committed:
DbSchema - simplify ifs while preserving readability

katzyn commented on h2/src/main/org/h2/command/dml/Optimizer.java:
```diff
@@ -100,9 +100,7 @@ private boolean canStop(int x) {
-            if (cost >= 0 && 10 * t > cost * TimeUnit.MILLISECONDS.toNanos(1)) {
-                return true;
-            }
+            return cost >= 0 && 10 * t > cost * TimeUnit.MILLISECONDS.toNanos(1);
```
This change actually generates worse and larger bytecode than it was before.

jlawrynowicz replied:
That's interesting. How about current version?

Also, shouldn't there be some comment about reason for checking ```x & 127```? Not sure why it's there.

katzyn replied:
If was expected, result of comparison is not a real `boolean` value in JVM so additional instructions are required.

```Java
a = <comparison>
```
is compiled like
```Java
if (<comparison>)
    a = true;
else
    a = false;
```
Current version saves 2 bytes but this current version is less readable.

katzyn replied:
I missed your question about condition.

`x & 127` is here to avoid execution of slower inner condition on each invocation. This is more or less oblivious.

katzyn commented on h2/src/main/org/h2/jdbc/JdbcBlob.java:
```diff
@@ -190,12 +190,11 @@ public OutputStream setBinaryStream(long pos) throws SQLException {
                 @Override
                 public void call() {
-                    value = c.createBlob(in, -1);
+                    value = conn.createBlob(in, -1);
```
This change generates an additional synthetic accessor method. Local final variable was used to avoid this.

jlawrynowicz committed:
Optimizer - simplify ifs

katzyn commented on h2/src/main/org/h2/util/ToDateParser.java:
```diff
@@ -309,15 +308,11 @@ public String toString() {
-                unmodifiedInputStr, unmodifiedFormatStr));
-        sb.append(format("\n      %s^%s ,  %s^ <-- Parsing failed at this point",
+        return format("\n    %s('%s', '%s')", functionName, unmodifiedInputStr, unmodifiedFormatStr) + format(
+                "\n      %s^%s ,  %s^ <-- Parsing failed at this point",
```
`+ format(` should be on another line.

katzyn commented on h2/src/main/org/h2/store/PageStore.java:
```diff
@@ -415,8 +415,8 @@ private void writeIndexRowCounts() {
-        for (int i = 0, size = list.size(); i < size; i++) {
-            writeBack(list.get(i));
+        for (CacheObject aList : list) {
+            writeBack(aList);
```
It's better not to use auto-generated names of variables.

jlawrynowicz committed:
JdbcBlob - revert changes to avoid synthetic method creation

jlawrynowicz committed:
DbSchema - revert inlining if/else

jlawrynowicz committed:
ToDateParser - formatting

jlawrynowicz committed:
improve variable naming

katzyn commented on h2/src/main/org/h2/jdbc/JdbcClob.java:
```diff
@@ -138,15 +138,14 @@ public Writer setCharacterStream(long pos) throws SQLException {
                 @Override
                 public void call() {
-                    value = c.createClob(IOUtils.getReader(in), -1);
+                    value = conn.createClob(IOUtils.getReader(in), -1);
```
The same synthetic method here.

katzyn commented on h2/src/main/org/h2/command/dml/Query.java:
```diff
@@ -307,11 +307,7 @@ private boolean sameResultAsLast(Session s, Value[] params,
-            return false;
-        }
-        return true;
+        return db.getModificationDataId() <= lastEval || getMaxDataModificationId() <= lastEval;
```
What you're trying to optimize here and in many similar places?

Comparison is not a boolean, &&, ¦¦, etc are not booleans too in bytecode. Compiler will add the same code anyway. There is no win. You're only making source code less consistent. If some additional conditions will be required their insertion may be more complicated too.

katzyn replied:
Compiler may even generate worse code in some cases with `return`, but it depends on used compiler.

jlawrynowicz replied:
It's not really an optimization, as readability improvement. If you find this shorter code harder to read, I'll just revert all those changes.

katzyn replied:
Such changes are OK when there is only one if-else operator that can be avoided at all.

But when there are additional `else if` blocks these changes only complicate reading and editing of the sources.

jlawrynowicz replied:
I see your point. Would you rather I reverted those changes or get rid of the if/else in those situations and add appropriate formatting?

katzyn replied:
This pull request will not be merged anyway. Send a new smaller requests.

In the new request you can change simple if-else (without else if) constructions.

But it's better not to touch large branches. Monster condition with a lot of || doesn't look good.

jlawrynowicz replied:
I've reverted those changes.

jlawrynowicz committed:
if inlining improvements, other review changes

grandinj commented:
git reflog is often useful in retrieving old commits

jlawrynowicz commented:
The PR is already split into 31 commits and review changes are applied. Please decide if you'll merge it as it is or I should open a separate PR for each change.

grandinj commented:
@katzen, with respect to the 8b40a4d601e338c1d3f2ff08e019081acfeca01a commit in this series,

at some point in the java VM series, this code:

     for (int i = 0, size = columns.size(); i < size; i++) {

was significantly faster than this code:

        for (Column c : columns) {

Do you know when that changed?

andreitokar commented:
I do not think it actually changed yet. In my JMH tests, indexed loop over ArrayList of 50 elements is still faster than Iterator-based one on java 1.8 by about 10%, but on java 9 performance is the same (not sure about garbage creation though). In any case, it is insignificant, given that code in question is not performance-critical. There is definitely a bigger fish to fry out there.
On the other hand, massive mechanical changes like this one, may present a hardship during merge without perceivable benefits,  for anyone who have outstanding changes on his branches.

katzyn commented:
@grandinj, formal situation with list iteration in Java is simple. There is a tagging interface `java.util.RandomAccess` from the Java 1.4 days. If some implementation of `java.util.List` is also implementing this empty interface it means that iteration of this list with `get(index)` should be faster in most cases than usage of `Iterator`.

This is true for `ArrayList` and `ArrayList` is instance of `RandomAccess`. But there are some other implementations where iterators are much faster and they are not tagged with this interface therefore.

If you don't know what kind of List is received iterators are safer, because even for lists that implement `RandomAccess` they are not very slower than index-based loop with `get(index)`. But for large linked lists etc. iterators are very much faster. Of course, `instanceof` may be used with two different branches for different collections, but in the most cases this is a bad idea.

In H2 codebese, however, in the most cases instances of `ArrayList` are used explicitly. But automated tools in IDEs usually don't care and convert all index-based or iterator-based loops to enhanced loops where this is possible. This is safe in performance terms for arrays, safe for all loops that were iterator-based already, safe for many collections, but not for array-based lists. The main idea of enhanced loops is to make code more readable, but not more faster. Also tools in IDEA are known for bad suggestions in performance terms for a long time, looks like they are syntax-based, but not implementation-based. Sometimes they got adjusted to be more realistic, like with `toArray(T[])`, but actually there are many places where without actual tests with JMH on different versions of Java and sometimes on different CPUs it's not easy to say which code will work faster.

Enhanced for loops will also block garbage collection of iterated collection until return of the method that use it (fixed in `javac` from Java 10, JDT from Eclipse is also affected). But this is not our case, H2 doesn't have code that actually needs it.

So hypothetically enhanced loops may be safely used in H2 for plain arrays and for collections that are not array-based. This makes code more readable and performance should be the same.

`ArrayList` (and some more rarely used collections) are different. May be it's better not to touch them.

katzyn commented:
@andreitokar

I run some additional tests on Java 9 and results are very mixed. So I used `hsdis` disassembler to check what's going on.

C2 on Java 9 inlines this iterator after a lot of executions. Main code is still larger because iterator checks `modCount`, for example, and have other differences. But there is no actual allocation of additional object. Performance should be still lower due to additional checks, but this difference may be not measurable even with JMH.

But in the most cases code will be simply interpreted as is. In this case index-based loop is faster for a very small lists and slower for a larger lists (about 15+ elements; this is strange). After a lot of executions code will be compiled and than recompiled a few of times. After initial compilation it looks like iterator version is very much faster, but after recompilation index-based version works slightly faster again.

So for modern JVM it looks like enhanced loop may be used for `ArrayList` too.

katzyn commented:
Custom tests are not stable because recompilation is not happen after fixed count of executions, it should be performed in parallel, but recompilation improves speed dramatically.

However, H2 still supports Java 7, and most such array lists have a very few elements, and many such methods are not executed often enough for JIT, so current index-based implementation still should be better in terms of performance. Even on Java 9 on such conditions index-based loop works faster.

From my point of view only readability and maintenance of the sources should be considered for decision about are such changes needed or not. But I don't have any specific opinion.
